<h1>Pre-procesamiento y limpieza de Datos</h1>

In [1]:
'''Importar la libreria para la realizacion de este laboratorio.'''
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt 

In [2]:
'''Obtener el Dataset para mostrar todo su contenido.'''
df = pd.read_csv("data_act_01.csv",sep=";")
df.sample(10)

CrimeId OriginalCrimeTypeName          OffenseDate CallTime  \
7730  160951719            Encampment  2016-04-04T00:00:00    12:31   
4091  160933183      Person Screaming  2016-04-02T00:00:00    19:46   
936   160921523  Wanted Vehicle / Sub  2016-04-01T00:00:00    12:11   
6318  160943050  Threats / Harassment  2016-04-03T00:00:00    19:04   
6618  160943582    Suspicious Vehicle  2016-04-03T00:00:00    22:25   
5088  160940941          Passing Call  2016-04-03T00:00:00     8:26   
9086  160960254                   915  2016-04-05T00:00:00     3:04   
2536  160930376    Suspicious Vehicle  2016-04-02T00:00:00     2:41   
6145  160942778       Muni Inspection  2016-04-03T00:00:00    17:42   
4483  160933911          Traffic Stop  2016-04-02T00:00:00    23:23   

             CallDateTime   Disposition                    Address  \
7730  2016-04-04T12:31:00           ADV  2100 Block Of Chestnut St   
4091  2016-04-02T19:46:00           UTL       200 Block Of Font Bl   
936   2016-04-01T12:11:00           REP      0 Block Of Webster St   
6318  2016-04-03T19:04:00           HAN      0 Block Of Maxwell Ct   
6618  2016-04-03T22:25:00           ADV   200 Block Of Rutledge St   
5088  2016-04-03T08:26:00           HAN     900 Block Of Market St   
9086  2016-04-05T03:04:00           GOA      2300 Block Of 16th St   
2536  2016-04-02T02:41:00           ADV       Wisconsin St/25th St   
6145  2016-04-03T17:42:00  Not recorded       Geary Bl/van Ness Av   
4483  2016-04-02T23:23:00           ADV           7th St/folsom St   

               City State AgencyId  Range      AddressType  
7730  San Francisco    CA        1    NaN  Premise Address  
4091  San Francisco    CA        1    NaN  Premise Address  
936   San Francisco    CA        1    NaN  Premise Address  
6318  San Francisco    CA        1    NaN  Premise Address  
6618  San Francisco    CA        1    NaN  Premise Address  
5088  San Francisco    CA        1    NaN  Common Location  
9086  San Francisco    CA        1    NaN  Premise Address  
2536  San Francisco    CA        1    NaN     Intersection  
6145  San Francisco    CA        1    NaN     Intersection  
4483  San Francisco    CA        1    NaN     Intersection

In [3]:
'''Observar el tipo de variables que contiene el dataframe'''
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10051 entries, 0 to 10050
Data columns (total 12 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   CrimeId                10051 non-null  int64  
 1   OriginalCrimeTypeName  10051 non-null  object 
 2   OffenseDate            10051 non-null  object 
 3   CallTime               10051 non-null  object 
 4   CallDateTime           10051 non-null  object 
 5   Disposition            10051 non-null  object 
 6   Address                10051 non-null  object 
 7   City                   9730 non-null   object 
 8   State                  10048 non-null  object 
 9   AgencyId               10051 non-null  object 
 10  Range                  0 non-null      float64
 11  AddressType            10051 non-null  object 
dtypes: float64(1), int64(1), object(10)
memory usage: 942.4+ KB


In [4]:
'''Detectar valores nulos y su conteo'''
df.isnull().sum()

CrimeId                      0
OriginalCrimeTypeName        0
OffenseDate                  0
CallTime                     0
CallDateTime                 0
Disposition                  0
Address                      0
City                       321
State                        3
AgencyId                     0
Range                    10051
AddressType                  0
dtype: int64

In [5]:
'''Observar la cantidad de filas en el dataset.'''
df.shape

(10051, 12)

# Primer error (Columnas Innecesarias)
Anlizando el dataset existen columnas que no aportan informacion valiosa para el desarrollo o analisis posterior de que sea util para la materia de mineria de datos. Es por esto que se decidio eliminar la columna Range y la columna CallDateTime. La columna Range no tenia ningun valor. Y la columna CallDateTime tenia informacion redundante. Corregimos el error de esta forma.

In [6]:
'''Eliminar la columna "Range" ya que observando los registros se observa que no posee ningun dato.'''
df = df.drop(['Range'], axis=1)
df.sample(10)

CrimeId OriginalCrimeTypeName          OffenseDate CallTime  \
7274  160950880                   909  2016-04-04T00:00:00     9:06   
934   160921516          Passing Call  2016-04-01T00:00:00    12:10   
2285  160924028    Homeless Complaint  2016-04-01T00:00:00    23:46   
1854  160923273            Trespasser  2016-04-01T00:00:00    19:50   
1345  160922369          Traffic Stop  2016-04-01T00:00:00    16:02   
7409  160951114                    Sw  2016-04-04T00:00:00    10:02   
3972  160932941       Fight No Weapon  2016-04-02T00:00:00    18:38   
6046  160942595                  500e  2016-04-03T00:00:00    16:53   
4568  160934040      Well Being Check  2016-04-02T00:00:00    23:57   
8960  160954060                22500e  2016-04-04T00:00:00    22:57   

             CallDateTime   Disposition                    Address  \
7274  2016-04-04T09:06:00           HAN      500 Block Of Minna St   
934   2016-04-01T12:10:00           HAN           Balboa St/8th Av   
2285  2016-04-01T23:46:00  Not recorded           9th St/natoma St   
1854  2016-04-01T19:50:00           GOA        200 Block Of 9th St   
1345  2016-04-01T16:02:00           ADV          Brannan St/4th St   
7409  2016-04-04T10:02:00           CIT  300 Block Of Edinburgh St   
3972  2016-04-02T18:38:00           UTL       0 Block Of Jessie St   
6046  2016-04-03T16:53:00           CAN       200 Block Of 22nd Av   
4568  2016-04-02T23:57:00           GOA      19th St/san Carlos St   
8960  2016-04-04T22:57:00           CIT     100 Block Of Girard St   

               City State AgencyId      AddressType  
7274  San Francisco    CA        1  Premise Address  
934   San Francisco    CA        1     Intersection  
2285  San Francisco    CA        1     Intersection  
1854  San Francisco    CA        1  Premise Address  
1345  San Francisco    CA        1     Intersection  
7409  San Francisco    CA        1  Premise Address  
3972  San Francisco    CA        1  Premise Address  
6046  San Francisco    CA        1  Premise Address  
4568  San Francisco    CA        1     Intersection  
8960  San Francisco    CA        1  Premise Address

In [7]:
'''Eliminar la columna "CallDateTime" ya que observando 
los registros es la misma informacion ya presentada en 
otras columas.'''
df = df.drop(['CallDateTime'], axis=1)
df.sample(10)

CrimeId OriginalCrimeTypeName          OffenseDate CallTime  \
3497  160932076   Silent Holdup Alarm  2016-04-02T00:00:00    14:44   
368   160920471      Well Being Check  2016-04-01T00:00:00     5:51   
9436  160961609              Sleepers  2016-04-05T00:00:00    12:06   
1909  160923376          Traffic Stop  2016-04-01T00:00:00    20:24   
7193  160950743                22500e  2016-04-04T00:00:00     8:28   
3483  160932055          Passing Call  2016-04-02T00:00:00    14:38   
5398  160941458          Traffic Stop  2016-04-03T00:00:00    11:14   
6688  160943697        Meet W/citizen  2016-04-03T00:00:00    23:08   
5480  160941592          Traffic Stop  2016-04-03T00:00:00    11:46   
9212  160960735                22500e  2016-04-05T00:00:00     8:27   

     Disposition                  Address           City State AgencyId  \
3497         NOM   400 Block Of Castro St  San Francisco    CA        1   
368           ND       Mission St/11th St  San Francisco    CA        1   
9436         ADV       Sansome St/bush St  San Francisco    CA        1   
1909         CIT   Harrison St/harriet St  San Francisco    CA        1   
7193         GOA     600 Block Of 22nd Av  San Francisco    CA        1   
3483         HAN        3rd St/quesada Av  San Francisco    CA        1   
5398         CIT       Bosworth St/elk St  San Francisco    CA        1   
6688          ND  1400 Block Of Hudson Av  San Francisco    CA        1   
5480         ADV     San Jose Av/baden St  San Francisco    CA        1   
9212         CIT      200 Block Of 2nd Av  San Francisco    CA        1   

          AddressType  
3497  Premise Address  
368      Intersection  
9436     Intersection  
1909     Intersection  
7193  Premise Address  
3483     Intersection  
5398     Intersection  
6688  Premise Address  
5480     Intersection  
9212  Premise Address

# Segundo error (Valores duplicados)

In [8]:
'''Conteo de ids duplicados, y verificacion de los mismos.'''
print("CrimeId: ",df['CrimeId'].duplicated().sum())
df[df.duplicated(subset=['CrimeId'],keep=False)];

CrimeId:  4


Como se aprecia, existen 4 registros repetidos los cuales analizamos y se tratan de 2 IDs que se registran en 3 ocasiones cada uno, de modo que el sistema asigna 4 repeticiones. Vamos a corregir este error.

In [9]:
'''Realizamos una función la cual crea un set de la lista
de los datos duplicados existentes en la columna CrimeId'''
duplic = df.index[df.duplicated(subset=['CrimeId'],keep=False)].tolist()
maximo = df['CrimeId'].max()

In [10]:
'''Mediante el uso de un ciclo for el cual debe 
recorrer la lista de duplicados. A cada duplicado
se le asigna un nuevo Id, el cual es el valor 
maximo de los IDs aumentado en 1.'''
for r in duplic:
  maximo += 1
  df['CrimeId'].iloc[r] = maximo

C:\Users\User\AppData\Local\Temp\ipykernel_18772\2075909998.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['CrimeId'].iloc[r] = maximo


In [11]:
'''Verificamos si se cambiaron los datos'''
dup = df[df['CrimeId'].duplicated(keep=False)]['CrimeId'].tolist()
dup

[]

# Tercer error (Rellenar valores faltantes)

In [12]:
'''Se reemplaza los valores vacios que existen 
en las columnas City por valores en Cero, 
ya que al tener valores nulos pueden generar
errores adicionales en este contexto.'''

df[['City', 'State']] = df[['City','State']].fillna(value=0)
df

CrimeId OriginalCrimeTypeName          OffenseDate CallTime  \
0      160903280     Assault / Battery  2016-03-30T00:00:00    18:42   
1      160912272    Homeless Complaint  2016-03-31T00:00:00    15:31   
2      160912590             Susp Info  2016-03-31T00:00:00    16:49   
3      160912801                Report  2016-03-31T00:00:00    17:38   
4      160912811                   594  2016-03-31T00:00:00    17:42   
...          ...                   ...                  ...      ...   
10046  160964210      Well Being Check  2016-04-05T00:00:00    23:38   
10047  160964216          Traffic Stop  2016-04-05T00:00:00    23:43   
10048  160964227          Traffic Stop  2016-04-05T00:00:00    23:48   
10049  160964229          Passing Call  2016-04-05T00:00:00    23:48   
10050  160964249          Traffic Stop  2016-04-05T00:00:00    23:54   

        Disposition                  Address           City State AgencyId  \
0               REP  100 Block Of Chilton Av  San Francisco    CA        1   
1               GOA  2300 Block Of Market St  San Francisco    CA        1   
2               GOA  2300 Block Of Market St  San Francisco    CA        1   
3               GOA      500 Block Of 7th St  San Francisco    CA        1   
4               REP       Beale St/bryant St  San Francisco    CA        1   
...             ...                      ...            ...   ...      ...   
10046           GOA    Hollister Av/hawes St  San Francisco    CA        1   
10047           ADV   California St/mason St  San Francisco    CA        1   
10048           CIT   Rosella Ct/onondaga Av  San Francisco    CA        1   
10049  Not recorded          0 Blk Herbst Rd  San Francisco    CA        1   
10050           ADV        Grant Av/geary St  San Francisco    CA        1   

           AddressType  
0      Premise Address  
1      Premise Address  
2      Premise Address  
3      Premise Address  
4         Intersection  
...                ...  
10046     Intersection  
10047     Intersection  
10048     Intersection  
10049     Geo-Override  
10050     Intersection  

[10051 rows x 10 columns]

In [13]:
'''Verificar la redundancia de datos en la columna de City'''
df.groupby('City').count()

CrimeId  OriginalCrimeTypeName  OffenseDate  CallTime  \
City                                                                   
0                  321                    321          321       321   
 S                   1                      1            1         1   
Brisbane             1                      1            1         1   
Daly City            5                      5            5         5   
Presidio             3                      3            3         3   
SAN FRANCISCO        1                      1            1         1   
San Francisco     9665                   9665         9665      9665   
Treasure Isla       51                     51           51        51   
Yerba Buena          3                      3            3         3   

               Disposition  Address  State  AgencyId  AddressType  
City                                                               
0                      321      321    321       321          321  
 S                       1        1      1         1            1  
Brisbane                 1        1      1         1            1  
Daly City                5        5      5         5            5  
Presidio                 3        3      3         3            3  
SAN FRANCISCO            1        1      1         1            1  
San Francisco         9665     9665   9665      9665         9665  
Treasure Isla           51       51     51        51           51  
Yerba Buena              3        3      3         3            3

<h1>Analisis</h1>
Observamos que existian 321 valores nulos que fueron reemplazados por 0 Analizando las direccciones de estas filas, nos podemos dar cuenta que todas pertenecen a la ciudad de San Francisco, asi que reemplazaremos estos Valores por San Francisco. Del mismo modo se observan dos errores, existe una columna con la letra S, probablemente de San Francisco, comprobaremos la direccion de esta para estar seguros de que la ciudad es San Francisco. Asi mismo observamos que existe un dato con mayusculas. Vamos a corregir ese error.

In [14]:
'''Se reemplaza el dato en mayusculas, por minusculas.'''
df['City'] = df['City'].replace('SAN FRANCISCO','San Francisco');

In [15]:
'''Se reemplaza ahora la letra S por la ciudad de San Francisco'''
df['City']= df['City'].replace(' S','San Francisco')

In [16]:
'''Se reemplaza ahora las filas de 0 por la ciudad de San Francisco'''
df['City']= df['City'].replace(0,'San Francisco')

In [17]:
'''Verificar la redundancia de datos en la columna de City otra vez'''
df.groupby('City').count()

CrimeId  OriginalCrimeTypeName  OffenseDate  CallTime  \
City                                                                   
Brisbane             1                      1            1         1   
Daly City            5                      5            5         5   
Presidio             3                      3            3         3   
San Francisco     9988                   9988         9988      9988   
Treasure Isla       51                     51           51        51   
Yerba Buena          3                      3            3         3   

               Disposition  Address  State  AgencyId  AddressType  
City                                                               
Brisbane                 1        1      1         1            1  
Daly City                5        5      5         5            5  
Presidio                 3        3      3         3            3  
San Francisco         9988     9988   9988      9988         9988  
Treasure Isla           51       51     51        51           51  
Yerba Buena              3        3      3         3            3

In [18]:
'''Verificamos nuevamente los valores nulos, 
como se observa existen 3 estados que son nulos.'''
df.isnull().sum()

CrimeId                  0
OriginalCrimeTypeName    0
OffenseDate              0
CallTime                 0
Disposition              0
Address                  0
City                     0
State                    0
AgencyId                 0
AddressType              0
dtype: int64

# Cuarto error (Valores desubicados)
Para este error, se tiene que la columna State posee algunos datos con "1" en lugar de "CA", haciendo referencia al 1 de su ID de agencia, esto quiere decir que los datos fueron recorridos un espacio hacia la derecha. para corregir este error realizaremos el siguiente cambio en el dataframe.

In [19]:
'''Se verifica la columna state para analizar los valores nulos.
Como se observa existen 3 valores que no tienen valor de estado,
asi que corregiremos este error reemplazando este valor por el estado
correspondiente que es CA.'''
df.groupby('State').count()

CrimeId  OriginalCrimeTypeName  OffenseDate  CallTime  Disposition  \
State                                                                       
0            3                      3            3         3            3   
CA       10048                  10048        10048     10048        10048   

       Address   City  AgencyId  AddressType  
State                                         
0            3      3         3            3  
CA       10048  10048     10048        10048

In [20]:
'''Reemplazo de los valores 0 de la columna estado.'''
df['State']= df['State'].replace(0,'CA')

In [21]:
'''Verificar nuevamente'''
df.groupby('State').count()

CrimeId  OriginalCrimeTypeName  OffenseDate  CallTime  Disposition  \
State                                                                       
CA       10051                  10051        10051     10051        10051   

       Address   City  AgencyId  AddressType  
State                                         
CA       10051  10051     10051        10051